### ISSUE TITLE
patient_data.history limited to 141 blood level readings

### ISSUE THREAD

**Issue by alfredoramirez3** [robberwick/pylibrelinkup#34]
patient_data.history only returns 141 items; not the complete history.

**Comment by robberwick** [comment]
`patient_data.history` should return the complete set of measurements that are reported by the LLU api's `/llu/connections/{patient_id}/graph` endpoint, which is not the full history of the patient's readings.

The raw response is exposed as the `raw` property on the `ConnectionResponse` model returned by the call to `read()`, so you should be able to check that nothing has been dropped.

what do you get if you run the following, replacing the `{email}`, `{password}`, and `{patient_id}` with your values?

```python
from pylibrelinkup import PyLibreLinkUp
import json

client = PyLibreLinkUp(email="{email}", password="{password}")
client.authenticate()

conn_response = client.read(patient_identifier="{patient_id}")
raw_data = json.loads(conn_response.raw)

print(f"patient data history length: {len(conn_response.history)}")
print(f"graph data length: {len(raw_data['graph_data'])}")
```

**Comment by robberwick** [comment]
interestingly, running that with known login details for me gives only 47 items, but they do match:

```
patient data history length: 47
graph data length: 47
```

**Comment by alfredoramirez3** [comment]
https://gist.github.com/khskekec/6c13ba01b10d3018d816706a32ae8ab2 
![image](https://github.com/user-attachments/assets/0e438d78-a91e-4393-b829-aeca88ef484c)
Please view comment by faysal-ali dated Feb 6 and ATouhou response dated Feb 10.. @faysal-ali
I was able to access more data than described in this Gist, using the "/logbook" endpoint. Same headers and setup as "/graph" url.
The data available from that is the same as is shown in the app under "log book / Logbog (danish)".

to your point, I was hoping to get "full history".
not sure if using the /graph endpoint provides "full history" or just more data points.

any chance of adding an additional method that utilizes the /logbook endpoint?

**Comment by alfredoramirez3** [comment]
sorry, correction ...

to your point, I was hoping to get "full history".
not sure if using the "/logbook" endpoint provides "full history" or just more data points.

any chance of adding an additional method that utilizes the "/logbook" endpoint?

**Comment by robberwick** [comment]
Hmm. looking at that thread, further down, there appears to be a suggestion that the data is only available from LibreView, and not LibreLinkUp (which would make sense given that the URL is `/glucoseHistory/history`, and doesn't have the usual `llu/connections/{patient_id}/{endpoint}` format.

I'll have a look into it though. If i can't implement that `glucoseHistory` endpoint via the LLU api, I'll add the `logbook` endpoint as a new method at the very least.

**Comment by alfredoramirez3** [comment]
Wonder why the api’s didn’t provide start and end timestamp arguments for data extraction.

I’m a retired Data Engineer/Data Integration. Poor API design.

**Comment by robberwick** [comment]
yeah, the API is _weird_ in places. don't even get me started on the fact that some of their json keys are camelCase and others are snake_case, even in the same object!

**Comment by robberwick** [comment]
hi @alfredoramirez3 - i have a [PR ](https://github.com/robberwick/pylibrelinkup/pull/35)to implement support for that logbook endpoint on a [working branch](https://github.com/robberwick/pylibrelinkup/tree/logbook-endpoint).  if you want to give it an early play and see how it works for you, you can uninstall the pypi installed version and install it directly from github.

```bash
pip uninstall pylibrelinkup
pip install git+https://github.com/robberwick/pylibrelinkup.git@logbook-endpoint
```

I still need to write up some docs for it, and i might end up fiddling with the `read` method now that it's not the only method for retrieving data, so it's a bit poorly named, but you should be able to have a play with that logbook endpoint and see how you get on.

**Comment by robberwick** [comment]
Hi @alfredoramirez3 - Support for the logbook endpoint has been merged and is included in the new [v0.6.0](https://github.com/robberwick/pylibrelinkup/releases/tag/v0.6.0) release. Note that this is not the `/glucoseHistory/history` endpoint you mentioned, as that endpoint is exclusive to LibreView and not available through the LibreLinkUp API used by this package.
